### ISSUE TITLE
dealii::Function versus lambda closures and std::function

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#4237]
One class I never really came to good terms with is the `dealii::Function` class. Especially the fact that it contains a huge interface (notion of time, derivatives, vectorized evaluation, etc).
The problem is that we almost never use this full interface.

As evidenced in the past I am a huge fan of lambda closures and the `std::function` object. So in my own code I implement problem data in the following way:
```
const auto g = [this](const Point<dim> &p) -> rank0_type {
  // ...
};
```
But this currently requires to wrap the object `g` into a `dealii::Function`:
```
const auto boundary_function = to_function(g);
dealii::VectorTools::interpolate_boundary_values(
    mapping, dof_handler, 0, boundary_function, constraint_matrix); 
```

So, what about providing some wrapper glue in deal.II itself to allow something like
```
dealii::VectorTools::interpolate_boundary_values(mapping, dof_handler, 0, 
  [this](const Point<dim> &p, unsigned int component) -> double {
    // ...
  }, constraint_matrix); 
```

I suggest that we try to solve this on the `dealii::Function` level (before we start to write thousands lines of code adding a function overload to everything). Easy and straight forward would be a wrapper function
```
to_function([this](const Point<dim> &p, unsigned int component) -> double { ... })
```
But maybe there is something smarter.

**Comment by tamiko** [comment]
Another genuine question: How often are these special derivative functions used at all? I have spotted one instance where `vector_gradient_list` is used in VectorTools (`exact_solution.vector_gradient_list (fe_values.get_quadrature_points()` line 6865).

It strikes me that from an application point of view it might be much nicer to simply require a plain function object, e.g.,
```
const auto gradient = [](const Point<dim> &p) -> rank1_type
```

For example, even our own derived classes like `AutoDerivativeFunction` or special functions only implement the gradient and nothing else.

**Comment by davydden** [comment]
I also use 'std::function' a lot and would really like if there would be no need to wrap this into 'dealii::Function' a for some simple cases like VectorTooks::project().

That said,

> rom an application point of view it might be much nicer to simply require a plain function object

I disagree here. In some cases it's really convenient to have a single object which implements everything (values, gradients, hessian -- all that for various components).

**Comment by bangerth** [comment]
Just as a point of reference, `to_function` is easily implemented using class `ScalarFunctionFromFunctionObject` and class `VectorFunctionFromScalarFunctionObject`. I like the idea of a function `to_function()`.

**Comment by tamiko** [comment]
>> from an application point of view it might be much nicer to simply require a plain function object
>
> I disagree here. In some cases it's really convenient to have a single
> object which implements everything (values, gradients, hessian -- all
> that for various components).

That's fine on user level. I am talking about the interfaces we have:

What I mean is that there is a huge difference between
  mysterious_function_1(const std::function<double(dealii::Point<dim>)> &);
and
  mysterious_function_2(const dealii::Function<dim> &);

For mysterious_function_1 you can infer from the type that this should
be the value of a scalar valued function. For mysterious_function_2 you
do not know what it uses (only values, gradients, or the hessian? Which
of the three different variants for querying values? at different time
points?)

**Comment by davydden** [comment]
> For mysterious_function_1 you can infer from the type that this should
be the value of a scalar valued function. For mysterious_function_2 you
do not know what it uses (only values, gradients, or the hessian?

that I agree. I am just saying that in some cases one needs `mysterious_function_2(const dealii::Function<dim> &);` and it's not convenient to do `mysterious_function_1(const std::function<double(dealii::Point<dim>)> &);`. 

Think of partition of unity method where you enriched the FE space with apriori knowledge about the solution. In this case what you need the enrichment function to provide (values, gradients or hessians, etc) is up to its usage in the weak form, whereas the FE itself, for simplicity, takes a dealii::Function as enrichment.
I will speak only for things that are relevant to my, but I am sure there are similar cases elsewhere.

My point is -- i like `std::function` and I use them a lot, but we shall not nuke `dealii::Function` as there are cases where one needs it.

**Comment by bangerth** [comment]
I think you need to start not from the interfaces, but for what they are used. You will find that `Function` objects are used in contexts where you don't know whether you have a scalar or vector-valued function (e.g., `VectorTools::integrate_difference`), where you don't know from the function's signature whether you need values or gradients (same function, depending on what norm you want to compute; sometimes you need *both* values and gradients), and you don't always know whether all or some of the components of a vector-valued function will be used. `dealii::Function` serves all of these purposes well. `std::function` does not. So `to_function` seems like a great idea to me, but wholesale deprecation of `dealii::Function` does not.

**Comment by davydden** [comment]
> So to_function seems like a great idea to me, but wholesale deprecation of dealii::Function does not.

fully agree, exactly what i think about the topic.

**Comment by tamiko** [comment]
OK. I read my statements now three times in a row. WHERE ON EARTH did I say something about deprecating `dealii::Function`?

**Comment by tamiko** [comment]
@bangerth:
I had a quick grep session over the library again. From the whole interface [1] the following methods are used internally in deal.II (excluding special function implementations):

 - vector_value, value_list: often used.
 - vector_gradient, gradient_list:   in `vector_tools.template.h` [2] and `manifold_lib.cc` [3]
 - vector_laplacian, laplacian_list: -- not used --
 - vector_hessian, laplacian_list:   -- not used --

Again, do we really need to pass a function object via `dealii::Function` if all that we do almost all the time is a call to `vector_value*`, or `*value_list`? More importantly, we default define this methods via a loop and a call to `value`, i.e. all speed advantages of filling a vector via one function call are likely nullified. [4]

Now, it is certainly a bit much to completely remove `dealii::Function` because this will break *all* user code in existence [5]. But we could discuss to for example

 - deprecate parts of the interface and introduce some modern C++11 style function signatures. (Thanks to perfect forwarding no runtime penalty involved).
 - Maybe even immediately removing parts of the interface? (Removing some virtual methods might be remarkably compatible with almost all user code) [6]

In addition it would be possible to
 - Make our function signatures lightweight. E.g. if `interpolate_boundary_values` only uses values, the function object should only contain this.
 - Maintain compatibility with `dealii::Function` by automatically converting `dealii::Function` to an appropriate type


In any way, I will very likely make a pull request in the near future for the following change:
 - generalize `Scalar/VectorFunctionFromFunctionObject` to arbitrary callable objects
 - write above `to_function` wrappers.


[1]
```
  virtual Number value (const Point<dim> &p, const unsigned int component = 0) const;
  virtual void vector_value (const Point<dim> &p, Vector<Number> &values) const;
  virtual void value_list (const std::vector<Point<dim>> &points, std::vector<Number> &values, const unsigned int component = 0) const;
  virtual void vector_value_list (const std::vector<Point<dim>> &points, std::vector<Vector<Number>> &values) const;
  virtual void vector_values (const std::vector<Point<dim>> &points, std::vector<std::vector<Number>> &values) const;

  virtual Tensor<1,dim, Number> gradient (const Point<dim> &p, const unsigned int  component = 0) const;
  virtual void vector_gradient (const Point<dim> &p, std::vector<Tensor<1,dim, Number>> &gradients) const;
  virtual void gradient_list (const std::vector<Point<dim>> &points, std::vector<Tensor<1,dim, Number>> &gradients, const unsigned int component = 0) const;
  virtual void vector_gradients (const std::vector<Point<dim>> &points, std::vector<std::vector<Tensor<1,dim, Number>>> &gradients) const;
  virtual void vector_gradient_list (const std::vector<Point<dim>> &points, std::vector<std::vector<Tensor<1,dim, Number>>> &gradients) const;

  virtual Number laplacian (const Point<dim> &p, const unsigned int component = 0) const;
  virtual void vector_laplacian (const Point<dim> &p, Vector<Number> &values) const;
  virtual void laplacian_list (const std::vector<Point<dim> > &points, std::vector<Number> &values, const unsigned int component = 0) const;
  virtual void vector_laplacian_list (const std::vector<Point<dim>> &points, std::vector<Vector<Number>> &values) const;

  virtual SymmetricTensor<2,dim,Number> hessian (const Point<dim> &p, const unsigned int component = 0) const;
  virtual void vector_hessian (const Point<dim> &p, std::vector<SymmetricTensor<2,dim,Number>> &values) const;
  virtual void hessian_list (const std::vector<Point<dim>> &points, std::vector<SymmetricTensor<2,dim,Number>> &values, const unsigned int component = 0) const;
  virtual void vector_hessian_list (const std::vector<Point<dim>> &points, std::vector<std::vector<SymmetricTensor<2,dim,Number> > > &values) const;
```

[2] three calls of the form `exact_solution.[...]`

[3] internal, implementational detail.

[4] The `FunctionTime` features are entirely unused within the library.

[5] I am also fully aware of that a change is only justified if the benefits outweigh the costs.

[6] We should also come to a conclusion how much of a change version 9.0 shall be. Shall we make a bit of a radical cut by fixing some design decisions, or not?

**Comment by davydden** [comment]
@tamiko I see a major flaw in your grepping: if something is not used within the library it does not mean that user's code/libraries are not using this functionality.

**Comment by davydden** [comment]
While we discuss `dealii::Function`, i would bring another issue which, in some scenarios, could be important: we always split evaluation of values/gradients/hessians. An example where this is not efficient is if a function is given in spherical coordinates. Then for each evaluation one does whole lot of similar actions (cartesion -> spherical) plus some minor differences. 

Along the same lines is grouping evaluation of various functions (MultiFunciton if you like). This is somewhat similar to `components`, but we may want to distinguish between the two.

Of course, in most real life examples such things do not take too much time (i.e. assembly of the RHS), but there are cases where a different interface could speed things up (i.e. `FE_Enriched`).

**Comment by bangerth** [comment]
If your goal is to shrink the interface (e.g., get rid of the `_list` and laplacian/hessian functions), then that's something I can support. But I do still think that `Function` is basically a class that does exactly what it's intended to do and that is not fundamentally broken.
